### PR TITLE
Problem: NetMQ cannot be used with async/await pattern

### PR DESCRIPTION
**Pull request by somdoron**
Solution: Add ReceiveAsync (and some overloads) to socket. Also, a NetMQRuntime must be created and run in order to signal the completion of the tasks

### CODE REVIEW

File: src/NetMQ/AsyncReceiveExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            return socket.ReceiveFrameStringAsync(SendReceiveConstants.DefaultEncoding);
+        }
+
+
+        /// <summary>
+        /// Receive a single frame from <paramref name="socket"/>, asynchronously, and decode as a string using <paramref name="encoding"/>.
+        /// </summary>
+        /// <param name="socket">The socket to receive from.</param>
+        /// <param name="encoding">The encoding used to convert the frame's data to a string.</param>
+        /// <returns>The content of the received message frame as a string and a  .</returns>
```
Comment (drewnoakes) [inline src/NetMQ/AsyncReceiveExtensions.cs]: Incomplete comment?

File: src/NetMQ/AsyncReceiveExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    [SuppressMessage("ReSharper", "UnusedMethodReturnValue.Global")]
+    public static class AsyncReceiveExtensions
+    {
+        #region Receiving a frame as a byte array
+
+        /// <summary>
+        /// Receive a single frame from <paramref name="socket"/>, asynchronously.
+        /// </summary>
+        /// <param name="socket">The socket to receive from.</param>
+        /// <param name="more"><c>true</c> , otherwise <c>false</c>.</param>
```
Comment (drewnoakes) [inline src/NetMQ/AsyncReceiveExtensions.cs]: This overload doesn't have a `more` parameter.

File: src/NetMQ/AsyncReceiveExtensions.cs (hunk 1 above)
Comment (somdoron) [reply src/NetMQ/AsyncReceiveExtensions.cs]: yes, fixing

**Comment by somdoron** [comment]
When core 3 is out we can also implement IAsyncEnumerable

File: src/NetMQ/AsyncReceiveExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            socket.AttachToRuntime();
+
+            var msg = new Msg();
+            msg.InitEmpty();
+
+            if (socket.TryReceive(ref msg, TimeSpan.Zero))
+            {
+                bool more = msg.HasMore;
+                msg.Close();
+                return Task.FromResult(more);
```
Comment (drewnoakes) [inline src/NetMQ/AsyncReceiveExtensions.cs]: Should probably cache a true result and a false result here to avoid allocating.

Comment (somdoron) [reply src/NetMQ/AsyncReceiveExtensions.cs]: addressed in https://github.com/zeromq/netmq/pull/804

**Pull request merged**